### PR TITLE
More primitive tests

### PR DESCRIPTION
**Pull request by BastienDegardins**
Add tests for primitives methods : 

- (1 primitiveAdd)
- (2 primitiveSubtract)
- (3 primitiveLessThan)
- (4 primitiveGreaterThan)
- (5 primitiveLessOrEqual)
- (6 primitiveGreaterOrEqual)
- (7 primitiveEqual)
- (8 primitiveNotEqual)
- (9 primitiveMultiply)

### CODE REVIEW

File: smalltalksrc/VMMakerTests/VMObjectLayoutTests.class.st [hunk 1]
```diff
@@ -61,6 +61,18 @@ VMObjectLayoutTests >> testHeaderOfObjectEncodesTheCorrectAmountOfSlots [
...
 
+{ #category : #'tests - integerValues' }
+VMObjectLayoutTests >> testNegativeIntegerValuesInRange [
+
+	self assert: (memory isIntegerValue: -17)
+]
+
+{ #category : #'tests - integerValues' }
+VMObjectLayoutTests >> testNegativeIntegerValuesNotInRange [
+
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMObjectLayoutTests.class.st:72]: would be nice to add a little comment such as "an integer smaller than the smallest negative integer is not a valid number"

File: smalltalksrc/VMMakerTests/VMObjectLayoutTests.class.st [hunk 2]
```diff
@@ -61,6 +61,18 @@ VMObjectLayoutTests >> testHeaderOfObjectEncodesTheCorrectAmountOfSlots [
 			equals: objSize ]
 ]
 
+{ #category : #'tests - integerValues' }
+VMObjectLayoutTests >> testNegativeIntegerValuesInRange [
+
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMObjectLayoutTests.class.st:66]: Why only -17?
Is is a magic number?
May be we could have -2, 0, -100
What about the smallest minimal negative integer?

File: smalltalksrc/VMMakerTests/VMObjectLayoutTests.class.st [hunk 3]
```diff
@@ -253,6 +265,30 @@ VMObjectLayoutTests >> testObjectWithMoreThan254SlotsHasTwoHeaders [
 		self assert: numSlots equals: slots ]
 ]
 
+{ #category : #'tests - integerValues' }
+VMObjectLayoutTests >> testPositiveIntegerBorderCase1 [
+"Test of the border case when int = 2r000111111 ... . Regarding the cCode implementation, the sign bit can be lost if this test fails. The maxCInteger and numSmallIntegerTagBits guaranties the portability of the test on 32 and 64 bit computer. "
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMObjectLayoutTests.class.st:270]: Please indent the comment.
What means border case, the max value possibly encoded on ...

guaranties -> guarantees

File: smalltalksrc/VMMakerTests/VMObjectLayoutTests.class.st [hunk 4]
```diff
@@ -253,6 +265,30 @@ VMObjectLayoutTests >> testObjectWithMoreThan254SlotsHasTwoHeaders [
...
+	self deny: (memory isIntegerValue: memory maxCInteger >> memory numSmallIntegerTagBits)
+]
+
+{ #category : #'tests - integerValues' }
+VMObjectLayoutTests >> testPositiveIntegerBorderCase2 [
+"Test of the border case when int = 2r111000000 ... . Regarding the cCode implementation, the sign bit can be lost if this test fails. The maxCInteger and numSmallIntegerTagBits guaranties the portability of the test on 32 and 64 bit computer. "
+	self deny: (memory isIntegerValue: ((memory maxCInteger >> memory numSmallIntegerTagBits) bitXor: (memory maxCInteger))) "<=> isIntegerValue: (0001111) Xor (111111 ...)"
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMObjectLayoutTests.class.st:277]: What means border case, the max value possibly encoded on ...

guaranties -> guarantees

May be using a not is more readable.

File: smalltalksrc/VMMakerTests/VMObjectLayoutTests.class.st [hunk 5]
```diff
@@ -253,6 +265,30 @@ VMObjectLayoutTests >> testObjectWithMoreThan254SlotsHasTwoHeaders [
...
+]
+
+{ #category : #'tests - integerValues' }
+VMObjectLayoutTests >> testPositiveIntegerValuesInRange [
+
+	self assert: (memory isIntegerValue: 17)
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMObjectLayoutTests.class.st:283]: why 17 only
why not largest small int too?

File: smalltalksrc/VMMakerTests/VMObjectLayoutTests.class.st [hunk 6]
```diff
@@ -253,6 +265,30 @@ VMObjectLayoutTests >> testObjectWithMoreThan254SlotsHasTwoHeaders [
...
+]
+
+{ #category : #'tests - integerValues' }
+VMObjectLayoutTests >> testPositiveIntegerValuesNotInRange [
+
+	self deny: (memory isIntegerValue: memory maxSmallInteger +1)
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMObjectLayoutTests.class.st:289]: space after +

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 7]
```diff
@@ -66,10 +66,53 @@ VMPrimitiveTest >> tearDown [
...
+
+{ #category : #'tests - primitiveAdd' }
+VMPrimitiveTest >> testPrimitiveAddFailsWithTypeErrorPreservesStack [
+	| string |
+	
+	string := self newString: '42'.
+	interpreter push: string. 
+	interpreter push: (memory integerObjectOf: 2).
+	
+
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:104]: remove one blank line

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 8]
```diff
@@ -66,10 +66,53 @@ VMPrimitiveTest >> tearDown [
...
+	
+]
+
+{ #category : #'tests - primitiveAdd' }
+VMPrimitiveTest >> testPrimitiveAddFailsWithTypeErrorOnSecondOperand [
+
+	interpreter push: (memory integerObjectOf: 2).
+	interpreter push: memory trueObject.
+	
+
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:88]: remove extra blank line

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 9]
```diff
@@ -92,31 +135,46 @@ VMPrimitiveTest >> testPrimitiveAddWithNoOverflowPopsOperands [
 	interpreter primitiveAdd. 
 	
 	interpreter popStack. 
+	"The previous lines pop the result of primiviteAdd (1 2) for us to check that 1 and 2 were poped from the Stack by checking that the next value is our marker, for instance 42"
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:138]: poped -> popped
+ Stack -> stack

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 10]
```diff
@@ -92,31 +135,46 @@ VMPrimitiveTest >> testPrimitiveAddWithNoOverflowPopsOperands [
...
 	
 	self assert: interpreter stackTop equals: (memory integerObjectOf: 42). 
 	
 	
 ]
 
-{ #category : #'tests - primitiveAtPut' }
+{ #category : #'tests - primitiveAdd' }
+VMPrimitiveTest >> testPrimitiveAddWithNoOverflowPreservesStack [
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:146]: testPrimitiveAddWithOverflowPreservesStack

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 11]
```diff
@@ -92,31 +135,46 @@ VMPrimitiveTest >> testPrimitiveAddWithNoOverflowPopsOperands [
...
+	self assert: (interpreter stackValue: 2) equals: (memory integerObjectOf: 16). 
+	
+	
+]
+
+{ #category : #'tests - primitiveAdd' }
 VMPrimitiveTest >> testPrimitiveAddWithOverflow [
 
-	interpreter push: (memory integerObjectOf: (memory maxSmallInteger)).
+	interpreter push: (memory integerObjectOf: (memory minSmallInteger)).
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:162]: No need for extra parenthese

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 12]
```diff
@@ -572,6 +630,169 @@ VMPrimitiveTest >> testPrimitiveAtPutPutsTheValueForAnIndexableObject [
...
+	
+	interpreter primitiveDivide. 
+	
+	self assert: interpreter stackTop equals: (memory integerObjectOf: 2). 
+	self assert: (interpreter stackValue: 1) equals: (memory booleanObjectOf: true).
+]
+
+{ #category : #'tests - primitiveDivide' }
+VMPrimitiveTest >> testPrimitiveDivideFailsOnFloatResult [
+	"The interpreter fails because 42%4 != 0. This seems intentionnal in the implementation"
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:659]: remov "This seems intentionnal."
May be explain the reason, may be  "Implement exact integer division, if division is not integer, fails"

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 13]
```diff
@@ -572,6 +630,169 @@ VMPrimitiveTest >> testPrimitiveAtPutPutsTheValueForAnIndexableObject [
...
+	
+	interpreter primitiveDivide. 
+	
+	self assert: interpreter failed.
+]
+
+{ #category : #'tests - primitiveDivide' }
+VMPrimitiveTest >> testPrimitiveDivideWithNegativeNumbers [
+	
+	interpreter push: (memory integerObjectOf: 4).
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:718]: Use 6 instead of 4 so that we can distinguish between -3 and -2

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 14]
```diff
@@ -572,6 +630,169 @@ VMPrimitiveTest >> testPrimitiveAtPutPutsTheValueForAnIndexableObject [
...
+
+{ #category : #'tests - primitiveEqual' }
+VMPrimitiveTest >> testPrimitiveEqualWithNoError1 [
+
+	interpreter push: (memory integerObjectOf: 16).
+	interpreter push: (memory integerObjectOf: 16).
+	
+	interpreter primitiveEqual.
+
+	self assert: interpreter stackTop equals: (memory booleanObjectOf: true).
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:745]: you can replace by trueObject and falseObject

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 15]
```diff
@@ -572,6 +630,169 @@ VMPrimitiveTest >> testPrimitiveAtPutPutsTheValueForAnIndexableObject [
...
+
+{ #category : #'tests - primitiveEqual' }
+VMPrimitiveTest >> testPrimitiveEqualWithNoError2 [
+
+	interpreter push: (memory integerObjectOf: 16).
+	interpreter push: (memory integerObjectOf: 42).
+	
+	interpreter primitiveEqual.
+
+	self assert: interpreter stackTop equals: (memory booleanObjectOf: false).
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:756]: same as above

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 16]
```diff
@@ -572,6 +630,169 @@ VMPrimitiveTest >> testPrimitiveAtPutPutsTheValueForAnIndexableObject [
...
+VMPrimitiveTest >> testPrimitiveEqualWithNoErrorPopsOperands [
+
+	interpreter push: (memory integerObjectOf: 42). "Marker"
+	interpreter push: (memory integerObjectOf: 16).
+	interpreter push: (memory integerObjectOf: 16).
+	
+	interpreter primitiveEqual. 
+	
+	interpreter popStack. 
+	"The previous lines pop the result of primiviteAdd (1 2) for us to check that 1 and 2 were poped from the Stack by checking that the next value is our marker, for instance 42"
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:769]: Same comment than before and this is not primitive add :)

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 17]
```diff
@@ -572,6 +630,169 @@ VMPrimitiveTest >> testPrimitiveAtPutPutsTheValueForAnIndexableObject [
...
+
+	self assert: interpreter stackTop equals: (memory integerObjectOf: 42). 
+]
+
+{ #category : #'tests - primitiveEqual' }
+VMPrimitiveTest >> testPrimitiveEqualWithTypeErrorOnFirstOperand [
+
+	interpreter push: (memory booleanObjectOf: true).
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:777]: trueObject

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 18]
```diff
@@ -572,6 +630,169 @@ VMPrimitiveTest >> testPrimitiveAtPutPutsTheValueForAnIndexableObject [
...
+	interpreter push: (memory integerObjectOf: 16).
+	
+	interpreter primitiveEqual.
+
+	self assert: interpreter failed. 
+]
+
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:784]: What about 

2 = 2.0 ?

It should fail.

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 19]
```diff
@@ -608,6 +829,146 @@ VMPrimitiveTest >> testPrimitiveGetImmutabilityReturnsTrueIfObjectIsImmutable [
...
+	interpreter primitiveGreaterOrEqual. 
+	
+	self assert: interpreter failed. 
+]
+
+{ #category : #'tests - primitiveGreaterOrEqual' }
+VMPrimitiveTest >> testPrimitiveGreaterOrEqualFailsWithTypeErrorOnSecondOperand [
+	
+	interpreter push: (memory integerObjectOf: 2).
+	interpreter push: (memory booleanObjectOf: true).
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:847]: trueObject ....
look for all the booleanObjectOf: ... and fix them :)

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 20]
```diff
@@ -608,6 +829,146 @@ VMPrimitiveTest >> testPrimitiveGetImmutabilityReturnsTrueIfObjectIsImmutable [
...
+{ #category : #'tests - primitiveGreaterOrEqual' }
+VMPrimitiveTest >> testPrimitiveGreaterOrEqualWithNoErrorEquals [
+	
+	interpreter push: (memory integerObjectOf: 0).
+	interpreter push: (memory integerObjectOf: 0).
+	
+	interpreter primitiveGreaterOrEqual. 
+
+	self assert: interpreter stackTop equals: (memory booleanObjectOf: true). 
+]
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:885]: What is 

1 <= 1.0

You can do the same for all comparisons

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 21]
```diff
@@ -608,6 +829,146 @@ VMPrimitiveTest >> testPrimitiveGetImmutabilityReturnsTrueIfObjectIsImmutable [
...
+VMPrimitiveTest >> testPrimitiveGreaterOrEqualWithNoErrorPopsOperands [
+	
+	interpreter push: (memory integerObjectOf: 42). "Marker"
+	interpreter push: (memory integerObjectOf: 1).
+	interpreter push: (memory integerObjectOf: 2).
+	
+	interpreter primitiveGreaterOrEqual. 
+	
+	interpreter popStack. 
+	"The previous lines pop the result of primiviteAdd (1 2) for us to check that 1 and 2 were poped from the Stack by checking that the next value is our marker, for instance 42"
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:897]: Same as above

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 22]
```diff
@@ -608,6 +829,146 @@ VMPrimitiveTest >> testPrimitiveGetImmutabilityReturnsTrueIfObjectIsImmutable [
...
+VMPrimitiveTest >> testPrimitiveGreaterThanWithNoErrorPopsOperands [
+	
+	interpreter push: (memory integerObjectOf: 42). "Marker"
+	interpreter push: (memory integerObjectOf: 1).
+	interpreter push: (memory integerObjectOf: 2).
+	
+	interpreter primitiveGreaterThan. 
+	
+	interpreter popStack. 
+	"The previous lines pop the result of primiviteAdd (1 2) for us to check that 1 and 2 were poped from the Stack by checking that the next value is our marker, for instance 42"
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:967]: copy paste!!!

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 23]
```diff
@@ -656,6 +1017,296 @@ VMPrimitiveTest >> testPrimitiveInstVarAtPutPutsTheValueForNonIndexable [
...
+VMPrimitiveTest >> testPrimitiveLessThanWithNoErrorPopsOperands [
+	
+	interpreter push: (memory integerObjectOf: 42). "Marker"
+	interpreter push: (memory integerObjectOf: 1).
+	interpreter push: (memory integerObjectOf: 2).
+	
+	interpreter primitiveLessThan. 
+	
+	interpreter popStack. 
+	"The previous lines pop the result of primiviteAdd (1 2) for us to check that 1 and 2 were poped from the Stack by checking that the next value is our marker, for instance 42"
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:1155]: copy paster

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 24]
```diff
@@ -656,6 +1017,296 @@ VMPrimitiveTest >> testPrimitiveInstVarAtPutPutsTheValueForNonIndexable [
...
+
+	self assert: interpreter stackTop equals: (memory integerObjectOf: 42). 
+]
+
+{ #category : #'tests - primitiveMod' }
+VMPrimitiveTest >> testPrimitiveMod [
+
+	interpreter push: (memory integerObjectOf: 16).
+	interpreter push: (memory integerObjectOf: 2).
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:1164]: Would be nice to have test with something else than power of 2?

We should also have tests with negative integers, ) and floats


10 \\ -3 
-10\\ 3

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 25]
```diff
@@ -656,6 +1017,296 @@ VMPrimitiveTest >> testPrimitiveInstVarAtPutPutsTheValueForNonIndexable [
...
+	interpreter push: (memory integerObjectOf: 1).
+	interpreter push: (memory integerObjectOf: 2).
+	
+	interpreter primitiveMultiply. 
+	
+	interpreter popStack. 
+	"The previous lines pop the result of primiviteAdd (1 2) for us to check that 1 and 2 were poped from the Stack by checking that the next value is our marker, for instance 42"
+	
+	self assert: interpreter stackTop equals: (memory integerObjectOf: 42). 
+]
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:1285]: would be nice to have a test with 0

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 26]
```diff
@@ -933,6 +1584,65 @@ VMPrimitiveTest >> testPrimitiveNewWithInvalidClassFails [
...
+VMPrimitiveTest >> testPrimitiveNotEqualWithNoErrorPopsOperands [
+
+	interpreter push: (memory integerObjectOf: 42). "Marker"
+	interpreter push: (memory integerObjectOf: 16).
+	interpreter push: (memory integerObjectOf: 18).
+	
+	interpreter primitiveNotEqual. 
+	
+	interpreter popStack. 
+	"The previous lines pop the result of primiviteAdd (1 2) for us to check that 1 and 2 were poped from the Stack by checking that the next value is our marker, for instance 42"
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:1619]: copy paste

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 27]
```diff
@@ -1239,3 +1949,109 @@ VMPrimitiveTest >> testPrimitiveStringAtPutShouldNotModifyStringIfFailedWhenNonC
...
+{ #category : #'tests - primitiveSubtract' }
+VMPrimitiveTest >> testPrimitiveSubtractFailsWithTypeErrorPreservesStack [
+
+	| string |
+	
+	string := self newString: '42'.
+	interpreter push: string. 
+	interpreter push: (memory integerObjectOf: 2).
+	
+
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:1988]: remove extra line

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 28]
```diff
@@ -1239,3 +1949,109 @@ VMPrimitiveTest >> testPrimitiveStringAtPutShouldNotModifyStringIfFailedWhenNonC
...
+VMPrimitiveTest >> testPrimitiveSubtractWithNoOverflowPopsOperands [
+
+	interpreter push: (memory integerObjectOf: 42). "Marker"
+	interpreter push: (memory integerObjectOf: 1).
+	interpreter push: (memory integerObjectOf: 2).
+	
+	interpreter primitiveSubtract. 
+	
+	interpreter popStack. 
+	"The previous lines pop the result of primiviteAdd (1 2) for us to check that 1 and 2 were poped from the Stack by checking that the next value is our marker, for instance 42"
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:2021]: copy paste

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 29]
```diff
@@ -1239,3 +1949,109 @@ VMPrimitiveTest >> testPrimitiveStringAtPutShouldNotModifyStringIfFailedWhenNonC
...
+	
+	self assert: interpreter stackTop equals: (memory integerObjectOf: 42). 
+	
+	
+]
+
+{ #category : #'tests - primitiveSubtract' }
+VMPrimitiveTest >> testPrimitiveSubtractWithOverflow [
+
+	interpreter push: (memory integerObjectOf: (memory minSmallInteger)).
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:2031]: unneedeed parentheses

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 30]
```diff
@@ -1239,3 +1949,109 @@ VMPrimitiveTest >> testPrimitiveStringAtPutShouldNotModifyStringIfFailedWhenNonC
...
+	interpreter push: (memory integerObjectOf: 2).
+	
+	interpreter primitiveSubtract. 
+	
+	self assert: interpreter failed.
+
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:2037]: spurious empty lines

File: smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st [hunk 31]
```diff
@@ -1239,3 +1949,109 @@ VMPrimitiveTest >> testPrimitiveStringAtPutShouldNotModifyStringIfFailedWhenNonC
...
+	minSmallInt:= (memory integerObjectOf: memory minSmallInteger). 
+	
+	interpreter push: minSmallInt. 
+	interpreter push: (memory integerObjectOf: 2).
+	
+	interpreter primitiveSubtract. 
+	
+	self assert: interpreter stackTop equals: (memory integerObjectOf: 2). 
+	self assert: (interpreter stackValue: 1) equals: minSmallInt.
+
```
Comment (Ducasse) [inline smalltalksrc/VMMakerTests/VMPrimitiveTest.class.st:2055]: No empty lines at the end
